### PR TITLE
Allow user to select individual TPU core to train on

### LINKED ISSUE #1539: Support training multiple models in parallel on each TPU core.
## 🚀 Feature
While training a model using K-Fold method, it would be beneficial to train each model parallelly on a separate TPU core. There should be a feature by which we can assign a model training process to a particular TPU core. Similar to `gpus=[0,2]`

### Motivation
I came across this [kernel](https://www.kaggle.com/abhishek/super-duper-fast-pytorch-tpu-kernel) by Abhishek Thakur, wherein he trains multiple ROBERT models in parallel on each TPU core. It trained the model fast by utilizing all cores. I have tried doing the same with Lightning but realized I cant select a TPU core with it.

### Pitch
Not very clear about it. Maybe inbuilt support for KFold method wherein the dataset is split accordingly and a model is trained on each TPU code in a Kfold manner. Or just the support to select a core and train on it. e.g `tpus=[1]`

### Additional context

Kernel by Abhishek Thakur: 
https://www.kaggle.com/abhishek/super-duper-fast-pytorch-tpu-kernel

### PR DESCRIPTION
**Pull request by lezwon**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?   
- [ ] Did you write any new necessary tests?  
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## What does this PR do?
Fixes #1539 

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by williamFalcon** [comment]
@lezwon there shouldn't be a new flag, it should behave like gpus...

And I guess we need to remove the num_tpu_cores arg and replace with the tpu_cores?
```python
Trainer(tpu_cores=1)
Trainer(tpu_cores=8)
Trainer(tpu_cores=[2])
Trainer(tpu_cores=[6])
```

@PyTorchLightning/core-contributors

**Comment by Borda** [comment]
> @lezwon there shouldn't be a new flag, it should behave like gpus...
> And I guess we need to remove the num_tpu_cores arg and replace with the tpu_cores?

 Do we really need a new flag? Can it be solved by generalized gpus... Rename gpus to something meaningful for both GPU / TPU and eventually CPU as we have also distributed cpu backend... What a out just `cores` as you cannot mix run on GPU - TPU in the same time...

**Comment by lezwon** [comment]
> @lezwon there shouldn't be a new flag, it should behave like gpus...
> 
> And I guess we need to remove the num_tpu_cores arg and replace with the tpu_cores?
> 
> ```python
> Trainer(tpu_cores=1)
> Trainer(tpu_cores=8)
> Trainer(tpu_cores=[2])
> Trainer(tpu_cores=[6])
> ```
> 
> @PyTorchLightning/core-contributors

@williamFalcon I suppose something like `Trainer(tpu_cores=[2])` would do. But `Trainer(tpu_cores=[2, 3])` would not work as this is not supported yet. Can we finalize the implementation by taking into account the suggestions given by @Borda ? I like the idea of having just `cores`.

**Comment by williamFalcon** [comment]
I think we need to be explicit about GPUs and TPUs... let's keep: 
```python
Trainer(gpus)
Trainer(tpu_cores=)
```

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -90,6 +90,7 @@ def __init__(
             gpus: Optional[Union[List[int], str, int]] = None,
             auto_select_gpus: bool = False,
             num_tpu_cores: Optional[int] = None,
+            tpu_id: Optional[int] = None,
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py:93]: so I can use only one TPU? not several with indexes like GPU?

Comment (lezwon) [reply pytorch_lightning/trainer/trainer.py:93]: Not as per my knowledge. xla_distributed only supports 1 or 8 cores. We can't selectively choose the cores. 
Ref: https://pytorch.org/xla/release/1.5/index.html#torch_xla.distributed.xla_multiprocessing.spawn

Comment (williamFalcon) [reply pytorch_lightning/trainer/trainer.py]: tpu_id is not needed...

Comment (lezwon) [reply pytorch_lightning/trainer/trainer.py]: @williamFalcon I have replaced it with `tpu_cores` as you suggested. Valid values are `1`/`8`/`[<1-8>]`

**Comment by lezwon** [comment]
@williamFalcon @Borda I need a review on this PR.

File: pytorch_lightning/trainer/distrib_parts.py [hunk 2]
```diff
@@ -498,7 +499,7 @@ def single_gpu_train(self, model):
 
     def tpu_train(self, tpu_core_idx, model):
         # put model on tpu
-        model.to(xm.xla_device())
+        model.to(xm.xla_device(self.tpu_id))
```
Comment (williamFalcon) [inline pytorch_lightning/trainer/distrib_parts.py:502]: i think this now makes it ONLY possible to train on 1 core no? not multiple cores

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -96,7 +96,7 @@ def __init__(
             num_processes: int = 1,
             gpus: Optional[Union[List[int], str, int]] = None,
             auto_select_gpus: bool = False,
-            num_tpu_cores: Optional[int] = None,
+            tpu_cores: Optional[Union[List[int], int]] = None,
```
Comment (williamFalcon) [inline pytorch_lightning/trainer/trainer.py:99]: we need to add deprecation warnings for the other arg no?

File: pytorch_lightning/trainer/distrib_parts.py (hunk 2 above)
Comment (Borda) [reply pytorch_lightning/trainer/distrib_parts.py:502]: I think so... @lezwon ^^

Comment (lezwon) [reply pytorch_lightning/trainer/distrib_parts.py:502]: I have noticed that if `self.tpu_id` is `None` and I use `xmp.spawn`, the model trains at the same speed it trains when all cores are being used. So I assumed that all cores are being used. I could add some logging to confirm. Or just add a conditional for `xm.xla_device()` maybe?

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (lezwon) [reply pytorch_lightning/trainer/trainer.py:99]: Gotcha! I'll keep the `num_tpu_cores` and add a  deprecation warning.

Comment (Borda) [reply pytorch_lightning/trainer/trainer.py:99]: if you keep `num_tpu_cores` no deprecation is needed

Comment (lezwon) [reply pytorch_lightning/trainer/trainer.py:99]: @Borda My idea is to replace `num_tpu_cores` with `tpu_cores` in the coming iterations. As @williamFalcon mentioned, I plan to support `num_tpu_cores` for now with a deprecation warning. Will that be fine?

Comment (Borda) [reply pytorch_lightning/trainer/trainer.py:99]: oh, I see now, sorry for the confusion... add deprecation for `num_tpu_cores`

Comment (lezwon) [reply pytorch_lightning/trainer/trainer.py:99]: cool :))

File: pytorch_lightning/trainer/__init__.py [hunk 4]
```diff
@@ -550,9 +550,10 @@ def on_train_end(self):
 
     Use `num_sanity_val_steps` instead. Will remove 0.8.0.
 
-num_tpu_cores
+tpu_cores
 ^^^^^^^^^^^^^
```
Comment (Borda) [inline pytorch_lightning/trainer/__init__.py:554]:
```suggestion
^^^^^^^^^
```

File: pytorch_lightning/trainer/trainer.py [hunk 5]
```diff
@@ -324,9 +324,11 @@ def __init__(
         self.on_gpu = True if (gpus and torch.cuda.is_available()) else False
 
         # tpu config
-        self.on_tpu = num_tpu_cores is not None
-        self.num_tpu_cores = num_tpu_cores
-        assert num_tpu_cores in [1, 8, None], 'num_tpu_cores can only be 1 or 8'
+        self.on_tpu = tpu_cores is not None
+        self.tpu_cores = tpu_cores
+        assert tpu_cores in [1, 8, None] or len(tpu_cores) == 1, 'tpu_cores can only be 1, 8 or [<1-8>]'
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py:329]: in docs is also `-1` co is allowed?

Comment (Borda) [inline pytorch_lightning/trainer/trainer.py:329]:
```suggestion
        assert self.tpu_cores in (1, 8, None) or (isinstance(self.tpu_cores, (list, tuple, set)) and len(self.tpu_cores) == 1), '`tpu_cores` can only be 1, 8 or [<1-8>]'
```

Comment (Borda) [reply pytorch_lightning/trainer/trainer.py:329]:
```
tpu = 2
tpu in (1, 8, None) or len(tpu) == 1
Traceback (most recent call last):
  File "/home/jb/.local/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2885, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-4-d1a5257bb743>", line 1, in <module>
    tpu in (1, 8, None) or len(tpu) == 1
TypeError: object of type 'int' has no len()
```

File: pytorch_lightning/trainer/trainer.py [hunk 6]
```diff
@@ -458,7 +460,7 @@ def __init__(
         # override dist backend when using tpus
         if self.on_tpu:
             self.init_tpu()
-            self.current_tpu_idx = None
+            self.current_tpu_idx = self.tpu_id
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py:463]: do we need this extra attribute `current_tpu_idx`?

File: pytorch_lightning/trainer/distrib_parts.py (hunk 2 above)
Comment (williamFalcon) [reply pytorch_lightning/trainer/distrib_parts.py:502]: ONLY when the user requests a specific TPU index should we use 
```model.to(xm.xla_device(self.tpu_id))``` otherwise, leave it as it was.

@Borda we need TPU tests to make sure this PR doesn't break functionality

File: pytorch_lightning/trainer/trainer.py (hunk 5 above)
Comment (lezwon) [reply pytorch_lightning/trainer/trainer.py:329]: I haven't come across any functionality that supports `-1` as value to `num_tpu_cores`

File: pytorch_lightning/trainer/trainer.py (hunk 6 above)
Comment (lezwon) [reply pytorch_lightning/trainer/trainer.py:463]: I haven't noticed it being used anywhere. I assumed it can be used for logging during training.

Comment (Borda) [reply pytorch_lightning/trainer/trainer.py:491]: I do not think it is needed

**Comment by Borda** [comment]
@lezwon pls rebase master as we have some test fixes there :]

File: pytorch_lightning/trainer/trainer.py (hunk 6 above)
Comment (lezwon) [reply pytorch_lightning/trainer/trainer.py]: Have omitted it :)

**Comment by lezwon** [comment]
@Borda I have made the requested changes. Need your review on it :]

**Comment by williamFalcon** [comment]
@Borda @lezwon this is great! let's merge?

**Comment by lezwon** [comment]
@williamFalcon sure.. Let's do it 👍

**Comment by Borda** [comment]
give me a sec to check it...

File: tests/test_deprecated.py [hunk 7]
```diff
@@ -96,6 +96,9 @@ def test_tbd_remove_in_v0_9_0_trainer():
         trainer = Trainer(progress_bar_refresh_rate=50, show_progress_bar=False)
     assert getattr(trainer, 'show_progress_bar')
 
+    with pytest.deprecated_call(match='v0.9.0'):
+        _ = Trainer(num_tpu_cores=8)
```
Comment (Borda) [inline tests/test_deprecated.py:100]:
```suggestion
        trainer = Trainer(num_tpu_cores=8)
        assert trainer.tpu_cores == 8
```

File: pytorch_lightning/trainer/trainer.py [hunk 8]
```diff
@@ -189,7 +189,10 @@ def __init__(
                 GPUs are configured to be in "exclusive mode", such
                 that only one process at a time can access them.
 
-            num_tpu_cores: How many TPU cores to train on (1 or 8).
+            tpu_cores: How many TPU cores to train on (1 or 8) / Single TPU to train on [1]
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py:192]: just 1 or 8, nothing in between?

File: pytorch_lightning/trainer/trainer.py [hunk 9]
```diff
@@ -342,9 +345,19 @@ def __init__(
...
-        self.on_tpu = num_tpu_cores is not None
-        self.num_tpu_cores = num_tpu_cores
-        assert num_tpu_cores in [1, 8, None], 'num_tpu_cores can only be 1 or 8'
+        if num_tpu_cores is not None:
+            rank_zero_warn("Argument `num_tpu_cores` is now set by `tpu_cores` since v0.7.6"
+                           " and this argument will be removed in v0.9.0", DeprecationWarning)
+
+        if tpu_cores is None:
+            tpu_cores = num_tpu_cores
+        self.on_tpu = tpu_cores is not None
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py:354]: this is not directly related to this PR, but it is not clear what will happen if a user sets `gpus` and `tpu_cores` in the same time

File: pytorch_lightning/trainer/trainer.py [hunk 10]
```diff
@@ -342,9 +345,19 @@ def __init__(
...
+        self.tpu_cores = tpu_cores
+        assert self.tpu_cores in (1, 8, None) or (
+            isinstance(self.tpu_cores, (list, tuple, set)) and len(self.tpu_cores) == 1
+        ), '`tpu_cores` can only be 1, 8 or [<1-8>]'
+
+        self.tpu_id = tpu_cores[0] if isinstance(tpu_cores, list) else None
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py:360]: so it can run just on single-core, if user defines [1,2,3] only `1` is used

File: tests/test_deprecated.py (hunk 7 above)
Comment (Borda) [reply tests/test_deprecated.py:100]: added in #1864

File: pytorch_lightning/trainer/trainer.py (hunk 8 above)
Comment (lezwon) [reply pytorch_lightning/trainer/trainer.py:192]: Yes. The `nprocs` argument for `xm.spawn` supports either 1 or max number of devices.
Source: https://pytorch.org/xla/release/1.5/index.html#torch_xla.distributed.xla_multiprocessing.spawn

File: pytorch_lightning/trainer/trainer.py (hunk 9 above)
Comment (lezwon) [reply pytorch_lightning/trainer/trainer.py:354]: I will check this out and provide an update :]

Comment (lezwon) [reply pytorch_lightning/trainer/trainer.py:354]: @Borda I ran this on kaggle:
`trainer = pl.Trainer(tpu_cores=[1], gpus=[2], precision=16, max_epochs=20)`
It threw an Exception:
```
MisconfigurationException: 
                You requested GPUs: [2]
                But your machine only has: []
```

File: pytorch_lightning/trainer/trainer.py (hunk 10 above)
Comment (lezwon) [reply pytorch_lightning/trainer/trainer.py:360]: @Borda it will throw an assertion error if the length of the list is greater than 1.

File: pytorch_lightning/trainer/trainer.py (hunk 9 above)
Comment (Borda) [reply pytorch_lightning/trainer/trainer.py:354]: I know, it was not relaying to this PR comment but more for us as conceptual think how to handle these configurations...
what would be you expected behaviour as a user if you set gpu and tpu?
- take any of them if available
- which has higher priority (to be used) if both are available
